<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Session information | SCE Gallery</title>
  <meta name="description" content="Session information | SCE Gallery" />
  <meta name="generator" content="bookdown 0.20 and GitBook 2.6.7" />

  <meta property="og:title" content="Session information | SCE Gallery" />
  <meta property="og:type" content="book" />
  
  <meta property="og:image" content="https://www.bioconductor.org/images/logo_bioconductor.gif" />
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Session information | SCE Gallery" />
  
  
  <meta name="twitter:image" content="https://www.bioconductor.org/images/logo_bioconductor.gif" />



<meta name="date" content="2020-08-26" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="favicon.ico" type="image/x-icon" />
<link rel="prev" href="plotting-reduced-dimensions.html"/>

<script src="libs/header-attrs-2.3/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Single-Cell Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="chapter" data-level="1" data-path="plotting-reduced-dimensions.html"><a href="plotting-reduced-dimensions.html"><i class="fa fa-check"></i><b>1</b> Plotting reduced dimensions</a>
<ul>
<li class="chapter" data-level="1.1" data-path="plotting-reduced-dimensions.html"><a href="plotting-reduced-dimensions.html#foreword"><i class="fa fa-check"></i><b>1.1</b> Foreword</a></li>
<li class="chapter" data-level="1.2" data-path="plotting-reduced-dimensions.html"><a href="plotting-reduced-dimensions.html#setting-up-the-data"><i class="fa fa-check"></i><b>1.2</b> Setting up the data</a></li>
<li class="chapter" data-level="1.3" data-path="plotting-reduced-dimensions.html"><a href="plotting-reduced-dimensions.html#visualizing-with-scater"><i class="fa fa-check"></i><b>1.3</b> Visualizing with <em><span>scater</span></em></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="session-information.html"><a href="session-information.html"><i class="fa fa-check"></i>Session information</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">SCE Gallery</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="session-information" class="section level1 unnumbered">
<h1>Session information</h1>
<button class="rebook-collapse">
View session info
</button>
<div class="rebook-content">
<pre><code>R version 4.0.0 Patched (2020-05-01 r78341)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.5 LTS

Matrix products: default
BLAS:   /home/luna/Software/R/R-4-0-branch-dev/lib/libRblas.so
LAPACK: /home/luna/Software/R/R-4-0-branch-dev/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    parallel  stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] bluster_0.99.1              scuttle_0.99.13            
 [3] scran_1.17.15               scater_1.17.4              
 [5] ggplot2_3.3.2               DropletUtils_1.9.10        
 [7] SingleCellExperiment_1.11.6 SummarizedExperiment_1.19.6
 [9] DelayedArray_0.15.7         matrixStats_0.56.0         
[11] Matrix_1.2-18               Biobase_2.49.0             
[13] GenomicRanges_1.41.6        GenomeInfoDb_1.25.10       
[15] IRanges_2.23.10             S4Vectors_0.27.12          
[17] BiocGenerics_0.35.4         DropletTestFiles_0.99.0    
[19] BiocStyle_2.17.0           

loaded via a namespace (and not attached):
 [1] ggbeeswarm_0.6.0              colorspace_1.4-1             
 [3] ellipsis_0.3.1                XVector_0.29.3               
 [5] BiocNeighbors_1.7.0           bit64_4.0.2                  
 [7] interactiveDisplayBase_1.27.5 AnnotationDbi_1.51.3         
 [9] codetools_0.2-16              R.methodsS3_1.8.0            
[11] knitr_1.29                    dbplyr_1.4.4                 
[13] R.oo_1.23.0                   rebook_0.99.5                
[15] graph_1.67.1                  shiny_1.5.0                  
[17] HDF5Array_1.17.3              BiocManager_1.30.10          
[19] compiler_4.0.0                httr_1.4.2                   
[21] dqrng_0.2.1                   assertthat_0.2.1             
[23] fastmap_1.0.1                 limma_3.45.10                
[25] later_1.1.0.1                 BiocSingular_1.5.0           
[27] htmltools_0.5.0               tools_4.0.0                  
[29] rsvd_1.0.3                    igraph_1.2.5                 
[31] gtable_0.3.0                  glue_1.4.1                   
[33] GenomeInfoDbData_1.2.3        dplyr_1.0.2                  
[35] rappdirs_0.3.1                Rcpp_1.0.5                   
[37] vctrs_0.3.2                   rhdf5filters_1.1.2           
[39] ExperimentHub_1.15.2          DelayedMatrixStats_1.11.1    
[41] xfun_0.16                     stringr_1.4.0                
[43] ps_1.3.4                      mime_0.9                     
[45] lifecycle_0.2.0               irlba_2.3.3                  
[47] statmod_1.4.34                XML_3.99-0.5                 
[49] AnnotationHub_2.21.3          edgeR_3.31.4                 
[51] zlibbioc_1.35.0               scales_1.1.1                 
[53] promises_1.1.1                rhdf5_2.33.7                 
[55] yaml_2.2.1                    curl_4.3                     
[57] memoise_1.1.0                 gridExtra_2.3                
[59] stringi_1.4.6                 RSQLite_2.2.0                
[61] BiocVersion_3.12.0            filelock_1.0.2               
[63] BiocParallel_1.23.2           rlang_0.4.7                  
[65] pkgconfig_2.0.3               bitops_1.0-6                 
[67] evaluate_0.14                 lattice_0.20-41              
[69] purrr_0.3.4                   Rhdf5lib_1.11.3              
[71] CodeDepends_0.6.5             bit_4.0.4                    
[73] tidyselect_1.1.0              processx_3.4.3               
[75] magrittr_1.5                  bookdown_0.20                
[77] R6_2.4.1                      generics_0.0.2               
[79] DBI_1.1.0                     pillar_1.4.6                 
[81] withr_2.2.0                   RCurl_1.98-1.2               
[83] tibble_3.0.3                  crayon_1.3.4                 
[85] BiocFileCache_1.13.1          rmarkdown_2.3                
[87] viridis_0.5.1                 locfit_1.5-9.4               
[89] grid_4.0.0                    blob_1.2.1                   
[91] callr_3.4.3                   digest_0.6.25                
[93] xtable_1.8-4                  httpuv_1.5.4                 
[95] R.utils_2.9.2                 munsell_0.5.0                
[97] beeswarm_0.2.3                viridisLite_0.3.0            
[99] vipor_0.4.5                  </code></pre>
</div>

<div id="refs" class="references hanging-indent">
<div>
<p>Amezquita, R. A., A. T. L. Lun, E. Becht, V. J. Carey, L. N. Carpp, L. Geistlinger, F. Marini, et al. 2020. “Orchestrating single-cell analysis with Bioconductor.” <em>Nat. Methods</em> 17 (2): 137–45.</p>
</div>
<div>
<p>McCarthy, D. J., K. R. Campbell, A. T. Lun, and Q. F. Wills. 2017. “Scater: pre-processing, quality control, normalization and visualization of single-cell RNA-seq data in R.” <em>Bioinformatics</em> 33 (8): 1179–86.</p>
</div>
<div>
<p>Zheng, G. X., J. M. Terry, P. Belgrader, P. Ryvkin, Z. W. Bent, R. Wilson, S. B. Ziraldo, et al. 2017. “Massively parallel digital transcriptional profiling of single cells.” <em>Nat Commun</em> 8 (January): 14049.</p>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="plotting-reduced-dimensions.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/Bioconductor/OrchestratingSingleCellAnalysis-base/edit/master/reduced-dimensions.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/Bioconductor/OrchestratingSingleCellAnalysis-base/commits/master/reduced-dimensions.Rmd",
"text": null
},
"view": {
"link": "https://github.com/Bioconductor/OrchestratingSingleCellAnalysis-base/blob/master/reduced-dimensions.Rmd",
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
